app/controllers/hashtags_controller.rb

```diff
@@ -1,7 +1,14 @@
 # frozen_string_literal: true
 
 class HashtagsController < ApplicationController
-  requires_login
+  # Anonymous users can still see public posts which may contain hashtags
+  requires_login except: [:by_ids]
+
+  def by_ids
+    raise Discourse::NotFound if SiteSetting.login_required? && !current_user
+
+    render json: HashtagAutocompleteService.new(guardian).find_by_ids(params)
```

Comment:
Generally, we want to avoid making the `params` object accessible outside of the controller since it can contain any user input. Instead, I think we should only permit the keys that are allowed and then pass it on as necessary. Also, I believe we are accepting ids which will eventually be used in a query. Do we need to limit the number of ids which we accept each time? We don't want to allow a user to trigger a query with 1000 ids for example.

Reply:
Yes good point; we have `HashtagAutocompleteService::HASHTAGS_PER_REQUEST` for the slug lookup and we could do the same for the ID lookup. `20` may be a little low though; maybe we  could double it?

Reply:
I realised `HashtagsController` also has no rate limit then, but we could add that in another PR. I don't think it's ever had one.